### ISSUE TITLE
CTRL+C delays, fails to kill

### ISSUE THREAD

**Issue by darkpixel** [uqee/sticky-cluster#3]
I'm wrapping an express app.  When I switch to sticky-cluster, everything starts and runs properly.

When I hit CTRL+C, I get `^C` printed to the screen and nothing happens.

After about 5 seconds I get dropped back to the prompt.  A `ps uax | grep node` shows one node process running.

I have to `kill -9` to stop the process.

With debugging enabled, I see:

```
^C[1465882313124] worker 1: kill
[1465882313133] worker 2: kill
[1465882313133] worker 3: kill
[1465882313133] worker 4: kill
[1465882313133] worker 5: kill
[1465882313133] worker 6: kill
[1465882313133] worker 7: kill
[1465882313133] worker 8: kill
[1465882313133] master: stopped
[1465882313134] worker 1: died
[1465882313134] worker 2: died
[1465882313135] worker 4: died
[1465882313135] worker 6: died
[1465882313135] worker 7: died
[1465882313138] worker 5: died
[1465882313138] worker 8: died
[1465882313140] worker 3: died
```

This is where it hangs for ~5 seconds before returning to the command prompt.

**Comment by uqee** [comment]
Hi, @darkpixel!

First of all, proxy server (master) does not close instantly, it actually exits only [when all connections are ended](https://nodejs.org/api/net.html#net_server_close_callback). Also workers themselves are being killed by a [SIGTERM signal](https://nodejs.org/api/process.html#process_process_kill_pid_signal), which does not guarantee an instant effect either. Probably these facts are the causes of your 5s delay.

Futhermore, your log shows that since you've pushed ^C everything actually stopped within less than a second.

As for the node process being active after all, it's interesting phenomena. Would it be convenient for you to provide a little more details about it (e.g. `ps` before the app was launched, at the time when it's working and also when it stopped)? Also, I've updated things a bit in a `fix/issue#3` branch, can you check it out and launch an example there?

**Comment by darkpixel** [comment]
In the initial report, I had it running under 'nodemon'.  I switched to running `nodejs bin/www` directly, and the output was identical, except it would hang at the end and never return.  CTRL+C would produce no output and do nothing.  I would have to kill it via `kill -9` or `CTRL+\`.

Installing from the fix/issue#3 branch gives slightly different results:

```
18:57:44 [aaron@ender:~/code/mspdna-receiver] [mspdna] master(+71/-54) 130 ± nodejs bin/www 
18:57:52.116  MASTER  start..
18:57:52.124  MASTER  ..started at port 8001
18:57:52.125  WORKERS  start..
18:57:52.130  WORKER #1  ..started
18:57:52.133  WORKER #2  ..started
18:57:52.135  WORKER #3  ..started
18:57:52.137  WORKER #4  ..started
18:57:52.141  WORKER #5  ..started
18:57:52.144  WORKER #6  ..started
18:57:52.154  WORKER #7  ..started
18:57:52.162  WORKER #8  ..started
```

At this point the app is up and working.  I hit `CTRL+C` and get the following output:

```
^C18:57:56.828  MASTER  stop..
18:57:56.830  MASTER  ..stopped
18:57:56.834  WORKERS  stop..
18:57:56.835  WORKER #1  stop..
18:57:56.835  WORKER #2  stop..
18:57:56.836  WORKER #3  stop..
18:57:56.836  WORKER #4  stop..
18:57:56.836  WORKER #5  stop..
18:57:56.836  WORKER #6  stop..
18:57:56.836  WORKER #7  stop..
18:57:56.836  WORKER #8  stop..
18:57:56.838  WORKER #1  ..stopped
18:57:56.840  WORKER #2  ..stopped
18:57:56.840  WORKER #3  ..stopped
18:57:56.840  WORKER #5  ..stopped
18:57:56.840  WORKER #4  ..stopped
18:57:56.841  WORKER #6  ..stopped
18:57:56.841  WORKER #8  ..stopped
18:57:56.843  WORKER #7  ..stopped
```

It hangs here and does nothing (waited about 5 minutes).  Hitting `CTRL+C` a few more times produces:

```
^C18:58:04.365  MASTER  stop..
[Error: Not running]
^C18:58:06.485  MASTER  stop..
[Error: Not running]
```

I have to `kill -9` it.

When it's not running:

```
19:02:40 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± ps uax | grep -i nodejs
aaron    19174  0.0  0.0  11752  2112 pts/26   S+   19:02   0:00 grep --color=auto -i nodejs
19:02:44 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± 
```

When it's running:

```
19:02:44 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± ps uax | grep -i nodejs
aaron    19315 23.0  0.0 1072600 31044 pts/25  Sl+  19:03   0:00 nodejs bin/www
aaron    19324 92.0  0.1 1016188 53488 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19329 91.0  0.1 1016196 53840 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19334 90.0  0.1 1147260 53672 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19339 94.0  0.1 1016188 53632 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19340 90.0  0.1 1081720 53520 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19345 92.0  0.1 1081724 53596 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19354 94.0  0.1 1081776 53296 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19355 92.0  0.1 1081720 53496 pts/25  Sl+  19:03   0:00 /usr/bin/nodejs /tank/code/mspdna-receiver/bin/www
aaron    19411  0.0  0.0  11752  2200 pts/26   R+   19:03   0:00 grep --color=auto -i nodejs
19:03:05 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± 
```

After `CTRL+C`:

```
19:03:05 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± ps uax | grep -i nodejs
aaron    19315  2.0  0.0 1072664 32796 pts/25  Sl+  19:03   0:00 nodejs bin/www
aaron    19514  0.0  0.0  11752  2216 pts/26   S+   19:03   0:00 grep --color=auto -i nodejs
19:03:17 [aaron@ender:/tank/code/mspdna-receiver] master(+71/-54) ± 
```

**Comment by darkpixel** [comment]
I'm going to dig through my code a bit more.  I added in sticky-cluster to an existing express app that was using clustering, and I wanted to add websocket support.  I'll see what I can find...

**Comment by uqee** [comment]
My first hunch: looks like it still has a non empty event loop, if there's no open connections than it probably needs to [unref](https://nodejs.org/api/timers.html#timers_unref) something. I'll return to work on it in the evening.

**Comment by uqee** [comment]
Please, try the updated `fix/issue#3` branch and tell whether it works for you. If no, I can add a forced shutdown, but I'd suggest to double check any remaining timers/connections (e.g. have a look at [this issue](https://github.com/tj/connect-redis/issues/133)) beforehand.

**Comment by darkpixel** [comment]
_facepalm_

Yeah, it was the Redis connection.

Thanks for the help.
